lib/rust/parser/src/macros/resolver.rs

```diff
@@ -237,6 +237,7 @@ impl<'s> Resolver<'s> {
                 self.context = Context::Statement;
             }
             token::Variant::BlockStart(_) => {
+                self.finish_current_line();
```

Comment:
This `finish_current_line` call fixes the `if_then_else_chained_block` test, but it breaks `then_block` and `else_block`. Do I need to _finish_ only when the macro is already saturated? What is the easiest way to find that out, @kazcw?

Reply:
We cannot do this unconditionally--in most cases a block *should* be attached to the preceding line.

We need to introduce a concept of *inline context*. When we enter a macro segment (by reading the header token), we enter inline context until the next newline. When the newline is reached, if any tokens were read while in inline context, then we `finish_current_line` and the following block will be attached to the full preceding macro invocation; in any other case, we don't call `finish_current_line`, allowing the child block to be attached to its preceding line as usual.

I could implement it after my current issue (#8683).

Reply:
> We need to introduce a concept of _inline context_. When we enter a macro segment (by reading the header token), we enter inline context until the next newline. 

In the next example we _enter a macro segment_ ...
```ruby
if True then
  False
```
...by seeing the `if` token...

> When the newline is reached, if any tokens were read while in inline context, 

...we have seen `if` token, some token(s) in between and the `then` token....

> then we `finish_current_line` and the following block will be attached to the full preceding macro invocation; 

...however calling `finish_current_line` would turn the example into `(if True then) False` which is not what we want. Or I am not getting the concept of _any tokens were read_ properly...

> > Do I need to finish only when the macro is already saturated?

...I still tend to think in the term of _saturation_. The `if True then` macro isn't saturated as we know there has to be something behind `then` - unsaturated macros would not `finish_current_line`.

> I could implement it after my current issue (https://github.com/enso-org/enso/issues/8683).

Maybe you will have to. However James asked the _engine-ers_ to learn the Rust parser part and I volunteered. The learning is valuable even if it doesn't deliver workable solution. Thanks for your hints.

Reply:
`then` is the macro segment header before the newline, so no tokens have been read in the segment before the newline occurs. (When the body of the `if` segment ends without a newline, we leave its inline context without incident).

I think there's only a semantic difference between our paradigms in one case: IIUC, in the saturation model, a macro segment cannot accept a block as its body unless the macro segment requires a body. In the inline-context model, an `everything()` pattern (which does match nothing) can match an expression on the same line, or a block, but not both.

Macros use `everything()` a lot; even when a segment is semantically required to be non-empty we don't generally use a pattern that must be non-empty, because a macro non-match results in interpreting the input as a non-macro. When we have a near miss, like a macro with a missing body in one segment, we allow it to match but wrap it in `Invalid` if necessary, because recognizing poorly-matched macros as such gives the best analysis and error reporting.

Maybe saturation is a better paradigm, but it's not obvious to me how to reconcile it with the use of optional segment bodies.

Reply:
> We cannot do this unconditionally

6f5ec2a8602deb6cc78d401b9ec5f02931ac9ac5 introduces `start_segment` flag and only finishes line when the flag isn't set. That fixes `if_then_else_chained_block` as well as `then_block` and `else_block` tests.

However there are still problems: 96 passed; 8 failed, failures: _attributes_in_types, doc_comments, type_constructors, type_def_defaults, type_def_full, type_def_nested, type_methods, type_operator_methods_.

I investigated `attributes_in_types` failure: 
```
  left: `"(BodyBlock #((ArgumentBlockApplication (TypeDef type A #() #()) #((Annotated \"@\" a (Ident z) #() ()) (Annotated \"@\" b () #() ()) (Ident x)))))"`,
 right: `"(BodyBlock #((TypeDef type A #() #((Annotated \"@\" a (Ident z) #(()) (Annotated \"@\" b () #(()) (Ident x)))))))"`
```
e.g. there is extra `ArgumentBlockApplication` in the tree. I am not sure how to differentiate this situation and avoid `finish_current_line`? Your guidance is needed, @kazcw. Thanks in advance for your help.

Btw. I am surprised there is a macro in `"type A\n @a z\n @b\n x"` at all! Do we treat `type` definition as macro?

Reply:
After looking in to this in more depth, I think the right solution is to define a new macro pattern type and use it in the if-then-else definition, so I've implemented it that way.